### ISSUE TITLE
OpenBSD support – exiting when selected video

### ISSUE THREAD

**Issue by EmTeeAge** [pystardust/ytfzf#327]
Hi!

I've just tried installing ytfzf on OpenBSD with all the dependencies it asked me to install (expect ueberzug, replaced it with chafa, but can't seem to get ytfzt to read my config). However, the main issue is that ytfzf exits after you press Enter during the video selection. 

Searching works fine and you get a list of videos, but when you attempt to press Enter/click or use any of the Alt+ shortcuts, ytfzf abruptly exits back into terminal. Is there anyone who could perhaps try it on their OpenBSD setup to see if it's working? Or perhaps what limitation could be causing that abrupt exit.

Thank you.

_If this is a nonsense issue, feel free to close it, I don't want to bother you_

**Comment by Euro20179** [comment]
>  can't seem to get ytfzt to read my config

The config should be in `$HOME/.config/ytfzf/conf.sh`, the config file in docs/conf.sh is just all default settings, so adding it will do nothing, unless you change it. you can also try `YTFZF_CONFIG_FILE="$HOME/.config/ytfzf/conf.sh" ytfzf ...` and see if that works.

> when you attempt to press Enter/click or use any of the Alt+ shortcuts, ytfzf abruptly exits back into terminal

Just in case the shell matters can you run `file /bin/sh` and put the output.

**Comment by Euro20179** [comment]
Try adding 
```sh
[ -z "$shortcuts" ] && return 0
```
on line `1066`

**Comment by EmTeeAge** [comment]
First of all, thank you for your suggestions.

Yes, I already had the config in ~/.config.ytfzf dir. I also set conf.sh to be executable. The reason why I think ytfzf doesn't pick up my config is because I have set
`thumb_disp_method="chafa"`
which is installed and works when called through terminal, however ytfzf with the -t option still complains that ueberzug was not found.
```shell 
Scraping Youtube...
Downloading Thumbnails...
ueberzug not found. Please install it.
```
Running `YTFZF_CONFIG_FILE="$HOME/.config/ytfzf/conf.sh" ytfzf ...` results in the same output as above.

`file /bin/sh` outputs
`/bin/sh ELF 32-bit LSB shared object, Intel 80386, version 1`

I've also tried adding the line and just like for you it did absolutely nothing :)

**Comment by EmTeeAge** [comment]
Okay, forget about the config issue. It reads it, because if I set
`show_thumbnails=0` it gets to the video list without complaining about thumbnails.

Howerver if I set it to `show_thumbnails=1` it say again that ueberzug was not found. For some reason it doesn't pick up chafa and expects ueberzug even with the config option `thumb_disp_method="chafa"`

**Comment by Euro20179** [comment]
I think I know the solution to your config problem, your config *is* working but ytfzf makes an unnecessary dependency check.

add a "#" to the start of `921`, that line is checking if you have ueberzug installed. you could also just remove the "ueberzug" string.

**Comment by Euro20179** [comment]
also I haven't been able to replicate your issue, I'm just trying to figure out where and why it might happen.

**Comment by EmTeeAge** [comment]
Maybe I've just somehow installed it wrong. I mean, OpenBSD isn't officially supported and with it's userbase size, I'm not sure if it's even viable, but I just installed OpenBSD on my laptop and wanted to give it a try. Perhaps I should have mentioned how I did it:

1. `git clone https://github.com/pystardust/ytfzf/` into home dir
2. Installed jq, fzf, chafa (mpv and youtube-dl already installed - I use yt-dlp, but binary is renamed to youtube-dl due to other software)
3. Moved ytfzf to /usr/bin, adjusted owner to root:bin and permissions to match the rest
4. Tried running ytfzf, moved config to home and played around a bit, but never got past the video selection screen

_Am I missing something?_

Dependency versions:
MPV
```shell 
mpv 0.33.1
FFmpeg library versions:
    libavutil            56.51.100
    libavacodec      58.91.100
    libavformat      58.45.100
    libswscale        5.7.100
    libavfilter         7.85.100
    libswresample 3.7.100
FFmpeg version 4.3.2
```
Youtube-DL (yt-dlp)
```shell 
2021.07.07
```
jq
```shell 
jq-1.6
```
fzf
```shell 
0.24 (devel)
```
Chafa
```shell 
Chafa version 1.6.0
```
I should also mention that I'm running dwm on OpenBSD 6.9 i386 on a pretty old laptop, but I'm not sure if any of that matters.

Perhaps OpenBSD just _does something_ differently? Unfortunately I'm not very experiened with programming or scripting yet, otherwise I would have already tried some code debugging.

PS: Your fix to comment out line `921` works great, but gets me to a new issue (I'll just briefly explain it here, because I don't want to flood the page with random OpenBSD errors that only happen to me):

On the left side, instead of the thumbnail, the following is shown:

```shell 
Chanell
Duration
Views
Date
Description  :
chafa: Error loading 'home/myuser/.cache/ytfzf/thumb/YSHZHTtbFbM.png : unable to open image '/home/marek/.cache/ytfzf/thumb/YSHZHTtbFbM.png' : No such file or directory @ error/blob.c/OpenBlob/2924
```

However, upon opening `/home/myuser/.cache/ytfzf/thumb` I am greeting by a lot of files including YSHZHTtbFbM.png, which can be totally opened using chafa and gets displayed in my terminal (even though it looks like from the 80s).

**Comment by Euro20179** [comment]
it should work on openbsd since this is just a shell script.

You installed it correctly (it can really be installed however you want since it's just 1 file aand a shell script).

Try installing rofi and running `YTFZF_EXTMENU="rofi -dmenu -fuzzy -width 1500" ytfzf -D search`

**Comment by Euro20179** [comment]
Try adding `cache_dir="/home/myuser/.cache"` in the config file.

I'm not sure what's wrong with the info, you could try updating to the unstable branch with `ytfzf --update-unstable`

**Comment by EmTeeAge** [comment]
Tried adding multiple different locations of cache_dir to config, including the one you suggested but without luck..
Also installed rofi and tried the line which unfortunately resulted in the same behavior, just with different menu screen. Updating to unstable (and commenting out the dependency check again) didn't help either...

I might try installing OpenBSD again in a VM and seeing if I can reproduce the same issue on another machine to make sure it isn't just some random thing in my laptop setup.

_I'm sorry because I feel like I'm wasting your time on this_

**Comment by Euro20179** [comment]
haha, you're not wasting my time, I'm not spending all my time on this,

> OpenBSD again in a VM

this is a good idea,

a couple of ideas in order from best to worst (only try one at a time):
1. try installing the normal youtube-dl `pip install youtube-dl` and uninstall yt-dlp `pip uninstall yt-dlp` or make sure youtube-dl is running youtube-dl and not yt-dlp.
    * this is the best idea i have currently because the output of yt-dlp -F is different from youtube-dl -F, and it looks for specific things and if they are not found, it will exit.
2. you could try commenting out lines `1639-1646` you could also try commenting line `1647`.
3. try using a different shell maybe `bash /usr/bin/ytfzf search` or dash if you have that installed

**Comment by EmTeeAge** [comment]
Okay, this was fun.

1. I had this in the back of my mind the whole time, but since most programs I've encountered treat youtube-dl/yt-dlc/yt-dlp interchangeably or don't allow the change, but still work when you rename the binary, I was like "that _surely_ cannot be the issue, _right...?_ aaand it wasn't. I've installed youtube-dl, replaced the binary, checked and still nothing.
2. Tried, didn't work.
3. Sometimes the worst ideas are the best. **After running ytfzf in bash, it finally got to the line `Opening Player: mpv [url]` and actually played the video.** I'm not very familiar with the differences between shells, but something about ksh makes it unable to run ytfzf properly (I would have expected it to not run at all or error out.)

I also tried to replace youtube-dl with yt-dlp again and it seems to be working with both just fine. Now I'm trying to run different flags (download, audio only etc.) to see if some of them causes any trouble, but so far so good.

Perhaps someone could add a line into readme mentioning OpenBSD, that it will probably work, but only under bash? Just in case someone wants to try this on their OpenBSD system in the future.

I should have thought of this, since it's just a shell script like you've mentioned and different shells and their features resulting in weird behavior in scripts are fairly common.

Thank you so much for help!

TLDR: Run ytfzf under bash e.g `bash ytfzf [search]`

**Comment by Euro20179** [comment]
> but something about ksh makes it unable to run ytfzf properly

interesting, I said it was the worst solution because i installed ksh and ran ytfzf with and without my config, and with the default config and it worked every time. maybe pacman's ksh is different than bsd ksh?
